### PR TITLE
Update scripts/setup_gemdaq.sh to work everywhere

### PR DESCRIPTION
**Pull request by lmoureaux**
Everywhere meaning lxplus, P5 and 904 machines, slc6 and cc7

Intermediate changes tracked at:
https://gist.github.com/lmoureaux/97c663d2ff63b977c5823a8e010c4e1d#file-setup-sh-L11

Changes wrt the latest Gist version:

* Restore zsh support

Related PR:
https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/109

### CODE REVIEW

File: scripts/setup_gemdaq.sh [hunk 1]
```diff
@@ -190,15 +270,16 @@ then
...
             else
-                echo wget https://github.com/cms-gem-daq-project/vfatqc-python-scripts/releases/download/v${VFATQC_VERSION}-dev${VFATQC_DEV_VERSION}/gempython_vfatqc-${VFATQC_VERSION}.tar.gz
-                wget https://github.com/cms-gem-daq-project/vfatqc-python-scripts/releases/download/v${VFATQC_VERSION}-dev${VFATQC_DEV_VERSION}/gempython_vfatqc-${VFATQC_VERSION}.tar.gz
+                echo $WGET https://github.com/cms-gem-daq-project/vfatqc-python-scripts/releases/download/v${VFATQC_VERSION}-dev${VFATQC_DEV_VERSION}/gempython_vfatqc-${VFATQC_VERSION}.tar.gz
+                $WGET https://github.com/cms-gem-daq-project/vfatqc-python-scripts/releases/download/v${VFATQC_VERSION}-dev${VFATQC_DEV_VERSION}/gempython_vfatqc-${VFATQC_VERSION}.tar.gz
             fi
         fi
-    echo python -m pip install gempython_vfatqc-${VFATQC_VERSION}.tar.gz
-    python -m pip install gempython_vfatqc-${VFATQC_VERSION}.tar.gz
+        ls >/dev/null # Forces a filesystem sync
```
Comment (bdorney) [inline scripts/setup_gemdaq.sh:282]: My own ignorance here, but what's this line for?

File: scripts/setup_gemdaq.sh [hunk 2]
```diff
@@ -207,41 +288,58 @@ fi
 
 # Setup locations
 ####################
-if [[ $SYSTEM_INFO == *"lxplus"* ]];
-then
-    export DATA_PATH=/afs/cern.ch/work/$USER[0,1]/$USER/CMS_GEM/Data/gemdata
-elif [[ $SYSTEM_INFO == *"gem904"* ]];
+if [[ -z $DATA_PATH ]]; # Don't override existing value
```
Comment (bdorney) [inline scripts/setup_gemdaq.sh]: Revert this change.  This requirement *is* problematic since if `DATA_PATH` is set by a user on a `904` machine or a `p5` machine additional necessary `env` variables will *not* be set.

If you want to implement this check to not over-write the `DATA_PATH` variable then do so *only* on `lxplus`.  For the P5 and 904 machines a user should not be using a different `DATA_PATH` location since data taken/analysis performed on those test stands/p5 should be placed in the centrally defined location given below. 

Additionally in the case of `lxplus` add an if statement that if the path does not exist it is made with a `mkdir -p $DATA_PATH` command.

File: scripts/setup_gemdaq.sh [hunk 3]
```diff
@@ -207,41 +288,58 @@ fi
...
+if [ ! -f $VENV_DIR/lib/python$PYTHON_VERSION/site-packages/gempython/gemplotting/mapping/shortChannelMap.txt ]
 then
     find $VENV_DIR/lib/python$PYTHON_VERSION/site-packages/gempython -type f -name buildMapFiles.py -exec python {} \;
 fi
+
+if [ ! -d "$DATA_PATH" ];
+then
+    echo "WARNING: The directory \"$DATA_PATH\" (\$DATA_PATH) doesn't exist."
+    echo "         Fix your \$DATA_PATH using export DATA_PATH=..."
+fi
```
Comment (bdorney) [inline scripts/setup_gemdaq.sh]: Remove lines 333-337.  See discussion [here](https://github.com/cms-gem-daq-project/sw_utils/pull/1/files#r193545436).  For 904 & p5 this will always exist and will be the centrally defined location; for `lxplus` it will be made if it does not exist.

**Review by bdorney — CHANGES_REQUESTED** [review]
Revision needed for post `venv` installation/activation env setup.

File: scripts/setup_gemdaq.sh (hunk 1 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:282]: At P5, since `wget` is executed remotely on `cmsusr`, the file appears on the local machine after some delay, and hence the installation fails. For whatever reason, `ls` appears to cause a synchronization, and the file is present afterwards. I'd be happy to find a cleaner solution.

File: scripts/setup_gemdaq.sh (hunk 2 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh]: > Additionally in the case of lxplus add an if statement that if the path does not exist it is made with a mkdir -p $DATA_PATH command.

I disagree. Rationale:

If `$DATA_PATH` doesn't exist when the user sets up its environment, I can see two cases:

1. Either the data is not where the user expects it to be (not downloaded yet, filesystem not mounted, erased...)
2. Or  `$DATA_PATH` is wrong (or not set yet).

In either case user intervention is required, so it's good to display a warning. Creating an empty directory will *not* solve any of the problems; worse, it will silence the warning.

File: scripts/setup_gemdaq.sh (hunk 1 above)
Comment (bdorney) [reply scripts/setup_gemdaq.sh:282]: If you can find a cleaner solution sure, but otherwise no problem.  Yes I've noticed the same behavior.

File: scripts/setup_gemdaq.sh (hunk 2 above)
Comment (bdorney) [reply scripts/setup_gemdaq.sh]: >Either the data is not where the user expects it to be (not downloaded yet, filesystem not mounted, erased...)

If the user hasn't downloaded it yet, this sets the *ideal* place on `AFS` where a `user` can navigate too and place the downloaded tarball.  

If the filesystem is not mounted or erased then there is a larger problem that's beyond our scope.  That would indicate that `AFS` isn't working on `lxplus`.  Recall I'm only asking for the directory creation in the case of `lxplus`.

> Or $DATA_PATH is wrong (or not set yet).

Again this would indicate a much wider problem (e.g. with `AFS` and `lxplus`) since the location which is hardcoded is always physically present in the `AFS` area. 

Please implement the creation of an empty directory.  This gives a `USER` a prime place to navigate too and place the data.

Comment (lmoureaux) [reply scripts/setup_gemdaq.sh]: > Please implement the creation of an empty directory. This gives a USER a prime place to navigate too and place the data.

Done. Will maintain a local patch not to have it in my own version.

File: scripts/setup_gemdaq.sh [hunk 4]
```diff
@@ -43,17 +58,20 @@ do
             VENV_ROOT="$OPTARG";;
         w)
             WIPE=1;;
+        P)
```
Comment (bdorney) [inline scripts/setup_gemdaq.sh:62]: add this to `helpstring` with an appropriate description.

File: scripts/setup_gemdaq.sh [hunk 5]
```diff
@@ -1,9 +1,22 @@
...
+then
+    # Not sourced from Bash
+    BASH_SOURCE="$0"
+    # For zsh (http://zsh.sourceforge.net/FAQ/zshfaq03.html#l18)
+    setopt shwordsplit || true
+fi
+
+helpstring="Usage: source $BASH_SOURCE [options]
     Options:
+        -p Path to the venv location
```
Comment (bdorney) [inline scripts/setup_gemdaq.sh]: and preferably keep this string alphabetised by argument

**Review by bdorney — DISMISSED** [review]

File: scripts/setup_gemdaq.sh (hunk 4 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:62]: It is there, isn't it? l#25

File: scripts/setup_gemdaq.sh [hunk 6]
```diff
@@ -106,6 +143,22 @@ else
   return 1
 fi
```
Comment (mexanick) [inline scripts/setup_gemdaq.sh:144]: Lines 112-144 would benefit from refactoring. Suggestion:
```bash
if [[ $SYSTEM_INFO == *"lxplus"* ]];
then
    # LCG 93 doesn't provide `virtualenv' in PATH
    VIRTUALENV="python -m virtualenv"
    PIP="python -m pip"
    if [[ $KERNEL_VERSION == *"2.6."* ]];
    then
        OS_VERSION="slc6"
        source /cvmfs/sft.cern.ch/lcg/views/LCG_93/x86_64-slc6-gcc7-opt/setup.sh
    elif [[ $KERNEL_VERSION == *"3.10."* ]];
    then
        OS_VERSION="cc7"
        source /cvmfs/sft.cern.ch/lcg/views/LCG_93/x86_64-centos7-gcc7-opt/setup.sh
    else
        echo "Unrecognized kernel version! Exiting..."
        return 1
    fi
elif [[ "$DNS_INFO" == "cms" ]];
then
    # We are in the .cms network
    WGET="ssh cmsusr wget"
 fi
```

File: scripts/setup_gemdaq.sh (hunk 1 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:282]: If you just need to introduce a delay, use a `sleep` command. E.g. if you want to pause for 0.5 second, you need `sleep 0.5`
Same comment in every place where you have `ls >/dev/null`

File: scripts/setup_gemdaq.sh [hunk 7]
```diff
@@ -209,7 +291,19 @@ fi
 ####################
 if [[ $SYSTEM_INFO == *"lxplus"* ]];
 then
-    export DATA_PATH=/afs/cern.ch/work/$USER[0,1]/$USER/CMS_GEM/Data/gemdata
+    if [[ -z "$DATA_PATH" ]]; # Don't override existing value
+    then
+        export DATA_PATH=/afs/cern.ch/work/${USER:0:1}/$USER/CMS_GEM/Data/gemdata
+    else
+        # Make sure it's exported
+        export DATA_PATH
```
Comment (mexanick) [inline scripts/setup_gemdaq.sh:300]: How this is supposed to work?

**Review by mexanick — CHANGES_REQUESTED** [review]

File: scripts/setup_gemdaq.sh (hunk 7 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:300]: Bash manual:
```
export [-fn] [name[=word]] ...
export -p
    The supplied names are marked for automatic export to the environment of subsequently executed commands.
```

File: scripts/setup_gemdaq.sh (hunk 4 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:62]: It was added to the helpstring as requested.

File: scripts/setup_gemdaq.sh (hunk 1 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:282]: Tried `sleep 1`, didn't work. `ls` is instant and works.

File: scripts/setup_gemdaq.sh (hunk 6 above)
Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:144]: You forgot the 904 machines, that can be either `slc6` or `cc7`

File: scripts/setup_gemdaq.sh (hunk 1 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:282]: What means it doesn't work? throws an error? or what?

Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:282]: > What means it doesn't work? throws an error? or what?

The downloaded file doesn't appear on the local machine within 1s.

File: scripts/setup_gemdaq.sh (hunk 6 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:144]: well, add it ;) Anyway, it should be refactored in a way that lines like 131-137 do not repeat.

Comment (mexanick) [reply scripts/setup_gemdaq.sh:144]: Just make the 
```bash
if [[ $KERNEL_VERSION == *"2.6."* ]];
...
```
the outmost condition

File: scripts/setup_gemdaq.sh (hunk 7 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:300]: But you do not assign anything to $DATA_PATH before...

Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:300]: If `$DATA_PATH` is empty or not defined, set and export it. Else, keep the original value and export it.

Comment (bdorney) [reply scripts/setup_gemdaq.sh:300]: > Else, keep the original value and export it.

If the else statement is triggered, it has already been exported, and so I think the point @mexanick is making is that this is unnecessary.

Comment (lmoureaux) [reply scripts/setup_gemdaq.sh:300]: Remember this file is to be sourced. Manual:
```
source filename [arguments]
    Read and execute commands from filename in the current shell environment
```
The key here is that the sourced file is executed in the *current shell*. Not in another process (normal execution), or even in a subshell. And hence all variables are accessible, including non-exported ones. The following examples demonstrate how the the three possible execution modes work.

**tl;dr**: A sourced script has access to all variables in the parent environment, and any variable set within a sourced script is accessible in the parent shell, which is why `source` is useful in the first place.

### Sourcing

Executed in the current shell and current process:
```
bash ~$ A=A
bash ~$ export B=B
bash ~$ echo 'echo $A $B; PI=3.14; export E=2.72' >test.sh
bash ~$ source test.sh
A B
bash ~$ echo $PI $E
3.14 2.72
```

### Subshell

Executed in another shell in the same process:
```
bash ~$ A=A
bash ~$ export B=B
bash ~$ (echo $A $B; PI=3.14; export E=2.72) # () creates a subshell
A B
bash ~$ echo $PI $E

```

### Executing

Executed in another shell in another process:
```
bash ~$ A=A
bash ~$ export B=B
bash ~$ echo 'echo $A $B; PI=3.14; export E=2.72' >test.sh
bash ~$ chmod +x test.sh
bash ~$ ./test.sh # Executed in another process, reads the shebang
B
bash ~$ echo $PI $E

```

**Comment by bdorney** [comment]
changed target branch from `master` to `develop`.

File: scripts/setup_gemdaq.sh (hunk 7 above)
Comment (mexanick) [reply scripts/setup_gemdaq.sh:300]: well, this would be really rare case to make problems) but as bulletproof conception is ok.

**Review by mexanick — APPROVED** [review]

**Review by bdorney — APPROVED** [review]
